src/curry/once.ts

```diff
@@ -0,0 +1,55 @@
+const onceResult = '__once__'
+
+/**
+ * The type of a function wrapped with `once`.
+ */
+export interface OnceFunction<
+  Args extends unknown[] = unknown[],
+  Return = unknown,
+  This = unknown,
+> {
+  (this: This, ...args: Args): Return
+  [onceResult]?: Return
+}
+
+/**
+ * Create a function that runs at most once, no matter how many times
+ * it's called. If it was already called before, returns the result
+ * from the first call. This is a lighter version of `memo()`.
+ *
+ * To allow the function to be called again, use `onceReset()`.
+ *
+ * ```ts
+ * const fn = once(() => Math.random())
+ * fn() // 0.5
+ * fn() // 0.5
+ * ```
+ */
+export function once<Args extends unknown[], Return, This = unknown>(
+  fn: (this: This, ...args: Args) => Return,
+): (this: This, ...args: Args) => Return {
+  const onceFn = function (this, ...args) {
+    if (!(onceResult in onceFn)) {
+      onceFn[onceResult] = fn.apply(this, args)
+    }
+    return onceFn[onceResult]
+  } as OnceFunction<Args, Return, This>
+  return onceFn
+}
+
+/**
+ * Reset the result of a function that was created with `once`,
+ * allowing it to be called again.
+ *
+ * ```ts
+ * const fn = once(() => Math.random())
+ * fn() // 0.5
+ * fn() // 0.5
+ * onceReset(fn)
+ * fn() // 0.3
+ * fn() // 0.3
+ * ```
+ */
+export function onceReset(fn: OnceFunction): void {
```

Comment:
Just an idea: we can put reset inside `once` 

```javascript
 const fn = once(() => Math.random())
 once.reset(fn)
```

It's maybe not so explicit. But `onceReset` can be confusing by itself.

Reply:
Hmmm, there's also the idea of defining `reset` as a method on the wrapper function itself. Not sure we have a guiding principle for this, but it could be said that `debounce` and `throttle` made a precedent for it.

Reply:
I'm a little wary of the method approach.  For example, you can accidentally call `reset` on a function which doesn't have it anymore (some type error in your code). Or loose a type and write something like `(fn as OnceFn).reset()`. With `reset(fn)` it looks cleaner and safer.